### PR TITLE
Switch to zero-to-nix installer

### LINKED ISSUE #273: Try to switch to zero-to-nix installer
We should consider switching to the https://zero-to-nix.com/start/install. They main advantage would be that it provides an easy way to uninstall nix.

Concerns:
1. It is collecting some telemetry: See https://github.com/DeterminateSystems/nix-installer#diagnostics.
2. It may not allow to add in a binary cache. (Update: It does allow that with `--extra-conf`.

### PR DESCRIPTION
**Pull request by soenkehahn**
Fixes https://github.com/garnix-io/garn/issues/273.

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -3,6 +3,7 @@ Vagrant.configure("2") do |config|
   config.vm.provider "virtualbox" do |vb|
     vb.memory = "4096"
   end
+  config.vm.network "private_network", ip: "192.168.56.5"
```
Comment (soenkehahn) [inline Vagrantfile:6]: Allows to access the test file server on the host.

File: flake.nix [hunk 2]
```diff
@@ -6,36 +6,94 @@
...
         #     testCommand = "hello";
         #   }
-        lib.mkInstallScript = { toolName, flakeLocation, testCommand }:
-          pkgs.writeScript
-            "installer"
-            (pkgs.lib.replaceStrings
-              [ "@@toolName@@" "@@flakeLocation@@" "@@testCommand@@" ]
-              [ toolName flakeLocation testCommand ]
-              (builtins.readFile ./install-template.sh));
+        lib.mkInstallScriptFiles = { toolName, baseUrl, allowHttp ? false, flakeLocation, testCommand }:
```
Comment (soenkehahn) [inline flake.nix:24]: `allowHttp` is necessary to be able to test this without ssl certs.

File: install-template.sh [hunk 3]
```diff
@@ -4,19 +4,25 @@
...
-  curl --proto '=https' --tlsv1.2 -sSfL https://releases.nixos.org/nix/nix-2.17.1/install -o "$TMP/install.sh"
-  chmod u+x "$TMP/install.sh"
+  curl @@forceHttpsOption@@ --tlsv1.2 -sSfL @@baseUrl@@/nix-installer.sh -o "$TMP/nix-installer.sh"
+  chmod u+x "$TMP/nix-installer.sh"
 
-  "$TMP/install.sh" --nix-extra-conf-file "$TMP/nix-extra-config" --daemon --yes
+  export NIX_INSTALLER_BINARY_ROOT=@@baseUrl@@
+  "$TMP/nix-installer.sh" \
+    install \
+    --no-confirm \
```
Comment (soenkehahn) [inline install-template.sh]: Do we want `--no-confirm` here? The zero-to-nix installer has much better output and asks much better questions. I'd say we don't.

File: install-template.sh [hunk 4]
```diff
@@ -28,11 +34,23 @@ if test_nix_installation; then
...
+https://github.com/DeterminateSystems/nix-installer. That installer
+will ask for your admin password to install nix.
+
+A note on uninstalling: One of the advantages of the Determinate Systems
+'nix-installer' is that it comes with an easy way to uninstall nix (and
+'@@toolName@@'):
+
+$ /nix/nix-installer uninstall
+
+EOF
```
Comment (soenkehahn) [inline install-template.sh:78]: Happy to change this section, if you have ideas/improvements!

File: install-template.sh [hunk 5]
```diff
@@ -28,11 +34,23 @@ if test_nix_installation; then
...
-  echo For more information, see: https://nixos.org/download
-  echo
+  cat << EOF
+Welcome to the '@@toolName@@' installer!
+
+'@@toolName@@' depends on nix, but it seems that you don't have a nix
+installation. This installer will install nix for you. It uses the
+'nix-installer' by Determinate Systems. For more information, see
+https://github.com/DeterminateSystems/nix-installer. That installer
+will ask for your admin password to install nix.
```
Comment (jkarni) [inline install-template.sh]: This last sentence is confusing: “that” seems to refer to “something else”.

Maybe have the sudo comment as a separate paragraph, and say “This” instead?

File: install-template.sh [hunk 6]
```diff
@@ -28,11 +34,23 @@ if test_nix_installation; then
...
+
+A note on uninstalling: One of the advantages of the Determinate Systems
```
Comment (jkarni) [inline install-template.sh]: I think just “If in the future you want to uninstall Nix, use:”

**Review by jkarni — APPROVED** [review]

**Pull request merged**